### PR TITLE
Member list order fixed for clients

### PR DESCRIPTION
**Pull request by sancar**
Member list order was broken because we used HashSet in protocol
related code generation part. ArrayList will be used in all
collections in code generation.
A test is added to verify memberlist of server and client is in
same order.

### CODE REVIEW

**Comment by sancar** [comment]
run-lab-run

**Comment by sancar** [comment]
run-lab-run

**Comment by sancar** [comment]
verify

File: hazelcast-client-legacy/src/test/java/com/hazelcast/client/cluster/ClientClusterServiceMemberListTest.java [hunk 1]
```diff
@@ -87,6 +89,13 @@ public void run()
         });
     }
 
+    @Test
+    public void testMemberListOrderConsistentWithServer() {
+        Set<Member> membersFromClient = client.getCluster().getMembers();
+        Set<Member> membersFromServer = dataInstance.getCluster().getMembers();
```
Comment (pveentjer) [inline hazelcast-client-legacy/src/test/java/com/hazelcast/client/cluster/ClientClusterServiceMemberListTest.java]: why do you cal the server instance a dataInstance? This is not very intuitive.

File: hazelcast-client/src/main/java/com/hazelcast/client/cache/impl/AbstractClientCacheProxy.java [hunk 2]
```diff
@@ -260,8 +260,8 @@ public V get(K key, ExpiryPolicy expiryPolicy) {
         Data expiryPolicyData = toData(expiryPolicy);
         ClientMessage request = CacheGetAllCodec.encodeRequest(nameWithPrefix, keySet, expiryPolicyData);
         ClientMessage responseMessage = invoke(request);
-        Set<Map.Entry<Data, Data>> entrySet = CacheGetAllCodec.decodeResponse(responseMessage).entrySet;
-        for (Map.Entry<Data, Data> dataEntry : entrySet) {
+        List<Map.Entry<Data, Data>> entries = CacheGetAllCodec.decodeResponse(responseMessage).entries;
+        for (Map.Entry<Data, Data> dataEntry : entries) {
```
Comment (pveentjer) [inline hazelcast-client/src/main/java/com/hazelcast/client/cache/impl/AbstractClientCacheProxy.java]: why do you call the collection entries, but a single item 'dataEntry'. So either uses entries/entry or dataEntries/dataEntry.

**Comment by pveentjer** [comment]
This PR contains a lot more than just fixing the ordering of the member listing. Also a rewrite was done for a lot of api's where sets were used and now replaced by lists and collections. This goes beyond what should happen during bug fixing period. 

I would suggest splitting the PR in 2 parts:
1: the bug fix for member ordering
2: the code rewriting; which can be merged in 3.7.

File: hazelcast-client-legacy/src/test/java/com/hazelcast/client/cluster/ClientClusterServiceMemberListTest.java (hunk 1 above)
Comment (sancar) [inline hazelcast-client-legacy/src/test/java/com/hazelcast/client/cluster/ClientClusterServiceMemberListTest.java]: I did not call it dataInstance. It was there before I wrote the test. I just used it.

**Comment by sancar** [comment]
Code rewriting part happened since we decided to remove the Set from Code Generation in protocol. And this decision came with the realizing it from member list response. Since we used HashSet wherever a Set is used in protocol, we lost the order that we have on server side. I don't know, if changing the original order causing a bug in other parts that Set is used.(sometimes performance is an issue when using HashSet as you know). So since it is a straightforward change, I decided to do all at once.

**Comment by asimarslan** [comment]
@pveentjer This PR contains client protocol related changes. Any required fix for the protocol cannot be delayed as such change directly broke the protocol compatibility. I know we do not make changes other than bug fixes. And I think this is a bug in protocol itself. Moreover postponing some part of this PR is not an option here due to compatibility concerns.

File: hazelcast-client-legacy/src/test/java/com/hazelcast/client/cluster/ClientClusterServiceMemberListTest.java (hunk 1 above)
Comment (pveentjer) [inline hazelcast-client-legacy/src/test/java/com/hazelcast/client/cluster/ClientClusterServiceMemberListTest.java]: Ok.

**Comment by pveentjer** [comment]
The arguments make sense. I'm 50% yes and no on this issue.

**Comment by asimarslan** [comment]
:+1:

**Comment by pveentjer** [comment]
I would like to move forward with this issue; so either get it merged or moved to 3.7. Can you make a ping at the architecture group and ask them for opinion.

**Comment by enesakar** [comment]
Peter, I also think arguments make sense. While both of us think in this direction no need to take architecture group's time. I approve the fixes.

**Pull request merged**